File: prepare-commit-msg/add-jira-to-commit-body.pl
#!/usr/bin/env perl

use strict;
use warnings;

require 5.008;

=head1 prepare-commit-hook

Searches your branch for a ticket of the form C<< LABEL >>-1234> and if it exists,
adds the ticket to the body of your commit when C<git commit> is invoked with
source C<message> or C<commit> (see L<https://git-scm.com/docs/githooks>, or
L<githooks(5)> for details).

For example:

    $ git checkout -b FOO-1234/my-new-feature
    # make some changes...
    $ git commit --all --message "Add the new feature!"
    $ git log -1
    commit 2c49333eabcd2f92d39a341ebaf47bbdb2e5b275 (HEAD -> FOO-1234/my-new-feature)
    Author: John Smith <john.smith@email.com>
    Date:   Mon Dec 7 20:25:45 2020 -0500

        Add the new feature!
        
        Ticket: FOO-1234

=cut

# EDIT ME: List of JIRA labels to accept
my @JIRA_LABELS = qw(FOO BAR);

# EDIT ME: Prefix for the ticket in the body
my $TICKET_PREFIX = "Ticket: ";

sub main {
    my ( $commit_msg_file, $commit_source, $commit_sha ) = @_;

    # Note: $commit_sha is unused

    my $branch_ticket = get_ticket_from_branch(@JIRA_LABELS) or return 0;
    add_ticket_to_commit( $branch_ticket, $commit_source, $commit_msg_file );
    return 0;
}

sub get_ticket_from_branch {
    my (@jira_labels) = @_;

    my $get_branch_cmd = 'git rev-parse --abbrev-ref HEAD';
    my $branch         = qx/$get_branch_cmd/;

    my $return_code = $? >> 8;
    if ( $return_code != 0 ) {
        warn "`$get_branch_cmd` exited with $return_code";
        return;
    }

    chomp($branch);

    for my $label (@JIRA_LABELS) {
        if ( $branch =~ m/(\Q$label\E-[0-9]+)/ ) {
            return $1;
        }
    }
    return;
}

sub get_ticket_string {
    my ($ticket) = @_;

    return "$TICKET_PREFIX $ticket";
}

sub add_ticket_to_commit {
    my ( $ticket, $commit_source, $commit_msg_file ) = @_;

    if ( is_source_supported($commit_source) ) {
        my $content = read_file_contents($commit_msg_file);
        write_ticket_to_file( $commit_source, $commit_msg_file, $content,
            $ticket );
    }
    return;
}

sub write_ticket_to_file {
    my ( $commit_source, $filename, $content, $ticket ) = @_;

    my $ticket_string = get_ticket_string($ticket);

    # noop if ticket already in commit
    return if $content =~ /\Q$ticket_string\E/;

    my $out_content = q//;
    if ( !$commit_source ) {
        $out_content = _content_for_new_commit( $content, $ticket_string );
    }
    elsif ( is_source_message($commit_source) ) {
        $out_content = _content_for_message( $content, $ticket_string );
    }
    elsif ( is_source_commit($commit_source) ) {
        $out_content =
          _content_for_exisiting_commit( $content, $ticket_string );
    }
    open( my $out, '>:encoding(utf8)', $filename )
      or die 'Could not open commit message file';
    print( $out $out_content );
    close($out) or warn 'Error closing commit file after write';
    return;
}

sub _content_for_new_commit {
    my ( $content, $ticket_string ) = @_;

    # Fresh commit, file contains some empty lines followed by the auto
    # generated comments, just insert our ticket before these comments
    return "\n\n" . $ticket_string . $content;
}

sub _content_for_message {
    my ( $content, $ticket_string ) = @_;

    # Commit file already contains the message, just insert the ticket after
    # this
    return $content . "\n\n" . $ticket_string;
}

sub _content_for_exisiting_commit {
    my ( $content, $ticket_string ) = @_;

    # Commit file already exists and holds a commit. We want to place the
    # ticket after the commit text, but before the auto generated comments
    return $content =~ s/(^#)/$ticket_string\n$1/mr;
}

sub read_file_contents {
    my ($filename) = @_;

    open( my $in, '<:encoding(utf8)', $filename )
      or die 'Could not open commit message file';
    local $/ = undef;    # 'slurp mode'
    my $content = <$in>;
    close($in) or warn 'Error closing commit file after read';
    return $content;
}

sub is_source_commit {
    my ($commit_source) = @_;

    # invoked with -c/--reuse-message or -C/--reedit-message or --ammed
    return $commit_source && $commit_source eq 'commit';
}

sub is_source_message {
    my ($commit_source) = @_;

    # invoked with --message/-m or -F/--file
    return $commit_source && $commit_source eq 'message';
}

sub is_source_supported {
    my ($commit_source) = @_;

    return
         !$commit_source
      || is_source_message($commit_source)
      || is_source_commit($commit_source);
}

exit main(@ARGV);

File: prepare-commit-msg/add-jira-to-commit-body.py
#!/usr/bin/env python3
"""
Hook to add a JIRA ticket from the current branch to the end of the commit body. You'll
want to:

* Edit `TICKET_LABELS` to support the labels used e.g. by JIRA
* Edit `TICKET_PREFIX` to what you want the ticket to be prefixed by in the commit
body

Searches your branch for a pattern of the form `<LABEL>-1234`, where `<LABEL>` is one of
the `TICKET_LABELS` below. It then adds the ticket to the commit if the commit source is
in the `SUPPORTED_SOURCES` below (see https://git-scm.com/docs/githooks or `githooks(5)`
for details on these sources).

For example:

        $ git checkout --branch FOO-1234/my-new-feature
        # make some changes...
        $ git commit --all --message "Add the new feature!"
        $ git log -1
        commit 2c49333eabcd2f92d39a341ebaf47bbdb2e5b275 (HEAD -> FOO-1234/my-new-feature)
        Author: John Smith <john.smith@email.com>
        Date:   Mon Dec 7 20:25:45 2020 -0500
            Add the new feature!

            Ticket: FOO-1234
"""

import re
import subprocess
import sys
from typing import Iterable, Literal, Optional, Sequence

TICKET_LABELS = ("FOO", "BAR")
TICKET_PREFIX = "Ticket: "
SUPPORTED_SOURCES = (None, "message", "template", "commit")

_AVAILABLE_SOURCES = (None, "message", "template", "merge", "squash", "commit")

# Returning nonzero would abort the commit
# I don't think there's a good reason to ever abort the commit
# just because of a missing ticket, so always return 0
def main(argv: Sequence[str]) -> Literal[0]:
    # see `githooks(5)` for details on expected args
    commit_msg_file = argv[1]

    commit_source: Optional[str]
    if len(argv) == 4:
        commit_source = argv[2]
    else:
        commit_source = None

    if commit_source in SUPPORTED_SOURCES:
        branch = _get_branch()
        if branch is not None:
            ticket = _get_ticket_from_branch(branch, TICKET_LABELS)
            if ticket is not None:
                with open(commit_msg_file) as f:
                    commit_content = f.read()
                out_content = _add_ticket_to_commit(
                    ticket, commit_source, commit_content
                )
                if out_content is not None:
                    with open(commit_msg_file, "w") as f:
                        f.write(out_content)

    return 0


def _add_ticket_to_commit(
    ticket: str, commit_source: Optional[str], commit_content: str
) -> Optional[str]:
    ticket_string = TICKET_PREFIX + ticket

    # no-op if the ticket details already exist
    if ticket_string in commit_content:
        return None

    if commit_source is None or commit_source in ("message", "commit"):
        return _add_ticket_details(commit_content, ticket_string, commit_source)
    else:
        return None


def _add_ticket_details(
    commit_content: str, ticket_string: str, commit_source: Optional[str]
) -> str:

    comment_start = re.search("^#", commit_content, flags=re.MULTILINE)
    if comment_start is not None:
        # commit open with editor: contents are the commit followed by
        # auto-generated comments
        # add the ticket between the end of the contents and the start of the comments
        comment_location = comment_start.span(0)[0]
        commit_text = commit_content[:comment_location]
        commit_comments = commit_content[comment_location:]

        ticket_string = f"\n{ticket_string}\n"
        if commit_source is None:
            ticket_string += "\n"

        return commit_text + ticket_string + commit_comments
    else:
        # commit not open with editor, just contains contents
        # append the ticket details
        return commit_content + f"\n{ticket_string}\n"


def _get_ticket_from_branch(branch: str, labels: Iterable[str]) -> Optional[str]:
    for label in labels:
        match = re.search(f"{re.escape(label)}-[0-9]+", branch)
        if match is not None:
            return match.group(0)
    return None


def _get_branch() -> Optional[str]:
    cmd = ("git", "rev-parse", "--abbrev-ref", "HEAD")
    try:
        proc = subprocess.Popen(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )
    except OSError as e:
        print(f"Failed to run `f{' '.join(cmd)}`: {e}", file=sys.stderr)
        return None

    stdout, stderr = proc.communicate()
    stdout_text = stdout.decode() if stdout is not None else ""
    stderr_text = stderr.decode() if stderr is not None else ""

    if proc.returncode != 0:
        print(
            f"`{' '.join(cmd)}` failed: {stdout_text}, {stderr_text}", file=sys.stderr
        )
        return None
    else:
        return stdout_text


if __name__ == "__main__":
    sys.exit(main(sys.argv))
